d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -5,109 +5,120 @@
 import com.linkedin.util.RateLimitedLogger;
 import com.linkedin.util.clock.SystemClock;
 import java.net.URI;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
 import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
-public class UriPropertiesDualReadMonitor {
+public class UriPropertiesDualReadMonitor
+{
   private static final Logger LOG = LoggerFactory.getLogger(UriPropertiesDualReadMonitor.class);
 
-  private final ConcurrentHashMap<String, ClusterMatchRecord> _clusters = new ConcurrentHashMap<>();
+  private final HashMap<String, ClusterMatchRecord> _clusters = new HashMap<>();
   // Limit error report logging to every 10 minutes
   private final RateLimitedLogger RATE_LIMITED_LOGGER =
       new RateLimitedLogger(LOG, TimeUnit.MINUTES.toMillis(10), SystemClock.instance());
   private int _totalUris = 0;
   private int _matchedUris = 0;
   private final DualReadLoadBalancerJmx _dualReadLoadBalancerJmx;
 
-  public UriPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx) {
+  public UriPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx)
+  {
     _dualReadLoadBalancerJmx = dualReadLoadBalancerJmx;
   }
 
-  public void reportData(String clusterName, UriProperties property, boolean fromNewLb) {
+  public synchronized void reportData(String clusterName, UriProperties property, boolean fromNewLb)
```

Comment:
synchronizing the whole method will not allow update on different clusters happen at the same time. For apps that read hundreds of clusters, this may slow down the thread.

We can just add synchronized setters for _totalUris and _matchedUris.

Reply:
well, since it's just for dual read, I guess it's fine.